torch/onnx/_internal/exporter/_core.py

```diff
@@ -116,15 +116,17 @@ def __init__(self, tensor: torch.Tensor, name: str | None = None):
     def numpy(self) -> npt.NDArray:
         self.raw: torch.Tensor
         if self.dtype == ir.DataType.BFLOAT16:
-            return self.raw.view(torch.uint16).numpy(force=True)
+            return (
+                self.raw.view(torch.uint16).numpy(force=True).view(self.dtype.numpy())
```

Comment:
Can you explain this workaround? Looks like it's getting around the issue that numpy does not support bfloat6?

Reply:
Sure! Since there is no bfloat16 in numpy, converting a tensor directly from torch would fail. Thus we view it as uint16 first in torch, get the numpy representation, and then re-view it with the ml_dtypes type we get from the onnx ir dtype.numpy() call to get an array ONNX IR expects (which has dtype ml_dtypes.bfloat16. )

Reply:
And uint16 is closer to bfloat16 in terms of the number of bit? Maybe worth mentioning somewhere.

Reply:
Will add a comment. Yes they are both 16 bit dtypes

Reply:
https://github.com/pytorch/pytorch/pull/151259